neonvm/controllers/virtualmachine_qmp_queries.go

```diff
@@ -246,57 +262,345 @@ func QmpQueryMemoryDevices(ip string, port int32) ([]QmpMemoryDevice, error) {
+	return err
+}
+
+func QmpDelMemoryBackend(mon *qmp.SocketMonitor, idx int) error {
+	cmd := []byte(fmt.Sprintf(
+		`{"execute": "object-del",
+		  "arguments": {"id": "memslot%d"}}`, idx,
+	))
+	_, err := mon.Run(cmd)
+	return err
+}
+
+func QmpAddMemoryDevice(mon *qmp.SocketMonitor, idx int) error {
+	cmd := []byte(fmt.Sprintf(
+		`{"execute": "device_add",
+		  "arguments": {"id": "dimm%d",
+						"driver": "pc-dimm",
+						"memdev": "memslot%d"}}`, idx, idx,
+	))
+	_, err := mon.Run(cmd)
+	return err
+}
+
+func QmpDelMemoryDevice(mon *qmp.SocketMonitor, idx int) error {
+	cmd := []byte(fmt.Sprintf(
+		`{"execute": "device_del",
+		  "arguments": {"id": "dimm%d"}}`, idx,
+	))
+	_, err := mon.Run(cmd)
+	return err
+}
+
+type QmpMemorySetter struct {
+	vm        *vmv1.VirtualMachine
+	targetCnt int
+	recorder  record.EventRecorder
+	log       logr.Logger
+
+	mon         *qmp.SocketMonitor
+	memBackends map[int]bool // idx -> is active
+	maxBackend  int          // to know where to start deletion
 
-	mon, err := QmpConnect(QmpAddr(virtualmachine))
+	memDevCount int
+	errs        []error
+}
+
+func (r *QmpMemorySetter) BuildState() error {
+	r.memBackends = map[int]bool{}
+	memDevs, err := QmpQueryMemoryDevices(QmpAddr(r.vm))
 	if err != nil {
 		return err
 	}
-	defer mon.Disconnect()
+	r.memDevCount = len(memDevs)
 
-	// try to find empty slot
-	var slot int32
-	for slot = 1; slot <= slots; slot++ {
-		cmd := []byte(fmt.Sprintf(`{"execute": "qom-list", "arguments": {"path": "/objects/memslot%d"}}`, slot))
-		_, err = mon.Run(cmd)
-		if err != nil {
-			// that mean such object wasn't found, or by other words slot is empty
-			break
+	for _, m := range memDevs {
+		idx, err := QmpMatchMemBackend(m.Data.Memdev)
+		if err == nil {
+			r.memBackends[idx] = true
 		}
 	}
 
-	// add memdev object
-	cmd := []byte(fmt.Sprintf(`{"execute": "object-add", "arguments": {"id": "memslot%d", "size": %d, "qom-type": "memory-backend-ram"}}`, slot, virtualmachine.Spec.Guest.MemorySlotSize.Value()))
-	_, err = mon.Run(cmd)
+	r.mon, err = QmpConnect(QmpAddr(r.vm))
 	if err != nil {
 		return err
 	}
-	// now add pc-dimm device
-	cmd = []byte(fmt.Sprintf(`{"execute": "device_add", "arguments": {"id": "dimm%d", "driver": "pc-dimm", "memdev": "memslot%d"}}`, slot, slot))
-	_, err = mon.Run(cmd)
+
+	backends, err := QmpQueryMemoryBackendIds(r.mon)
 	if err != nil {
-		// device_add command failed... so try remove object that we just created
-		cmd = []byte(fmt.Sprintf(`{"execute": "object-del", "arguments": {"id": "memslot%d"}}`, slot))
-		mon.Run(cmd)
 		return err
 	}
 
+	for b, _ := range backends {
+		if _, ok := r.memBackends[b]; !ok {
+			r.memBackends[b] = false
+		}
+	}
+	for idx, _ := range r.memBackends {
+		r.maxBackend = max(r.maxBackend, idx)
+	}
+	r.log.Info("QMP memory state", "backends", r.memBackends, "maxBackend", r.maxBackend)
 	return nil
 }
 
+func (r *QmpMemorySetter) Disconnect() {
+	if r.mon != nil {
+		r.mon.Disconnect()
```

Comment:
The error returned from `r.mon.Disconnect` is ignored. Probably you want to return it here.